### PR TITLE
8329995: Restricted access to `/proc` can cause JFR initialization to crash

### PR DESCRIPTION
**Pull request by jbachorik**
Please, review this trivial change to make using `ProcessIterator` more robust in the presence of SELinux or a similar system.

The call to `os::opendir("/proc")` may return `nulltptr` if the `/proc` is not accessible due to restrictions placed by the SELinux. In that case the `ProcessIterator` will SIGSEG because it assumes the `_dir`, which is the variable storing the result  of the `os::opendir("/proc")` call to be non-null.

The patch is missing regression test because it is very hard to simulate `/proc` not being accessible to the test process.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329995](https://bugs.openjdk.org/browse/JDK-8329995): Restricted access to `/proc` can cause JFR initialization to crash (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Erik Gahlin](https://openjdk.org/census#egahlin) (@egahlin - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18775/head:pull/18775` \
`$ git checkout pull/18775`

Update a local copy of the PR: \
`$ git checkout pull/18775` \
`$ git pull https://git.openjdk.org/jdk.git pull/18775/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18775`

View PR using the GUI difftool: \
`$ git pr show -t 18775`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18775.diff">https://git.openjdk.org/jdk/pull/18775.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18775#issuecomment-2053974163)

### CODE REVIEW

File: src/hotspot/os/linux/os_perf_linux.cpp [hunk 1]
```diff
@@ -847,7 +847,7 @@ SystemProcessInterface::SystemProcesses::ProcessIterator::ProcessIterator() {
 bool SystemProcessInterface::SystemProcesses::ProcessIterator::initialize() {
   _dir = os::opendir("/proc");
   _entry = nullptr;
-  _valid = true;
+  _valid = _dir != nullptr;
   next_process();
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_perf_linux.cpp:851]: A comment on the opendir would be useful e.g.
```
 _dir = os::opendir("/proc");  // May be null if /proc is not accessible.
```

**Review by dholmes-ora — COMMENTED** [review]
This change doesn't seem sufficient as many of the ProcessIterator functions access something under /proc without checking is_valid. Is it just  direct access to /proc that fails rather than specific entries?

**Comment by jbachorik** [comment]
@dholmes-ora  The other places are either checking for `null` result or for error code. I re-checked them and they are not vulnerable to this problem. I agree that the access might be more consistent but I didn't want to introduce any large sweeping changes just for the sakes of the changes.

**Review by egahlin — APPROVED** [review]

**Comment by jbachorik** [comment]
@dholmes-ora Are you fine with my explanation of the situation?

**Review by dholmes-ora — APPROVED** [review]
@jbachorik  - apologies I have been away a bit lately.

I can't quite convince myself that we may not have problems accessing /proc elsewhere because in a few places the path is constructed in one part of the code and then used somewhere else. But that is not the concern of this PR which is fixing a specific observed problem.

Thanks

**Comment by jbachorik** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**